token/program-2022/src/offchain.rs

```diff
@@ -46,14 +61,1005 @@ where
+            amount,
+        )
+        .await
+        .unwrap();
+
+        // Then use the offchain function to add the extra account metas to the
+        // _execute_ instruction from offchain
+        let mut offchain_execute_instruction = spl_transfer_hook_interface::instruction::execute(
+            &transfer_hook_program_id,
+            &source_pubkey,
+            &mint_pubkey,
+            &destination_pubkey,
+            &authority_pubkey,
+            &validate_state_pubkey,
+            amount,
+        );
+
+        ExtraAccountMetaList::add_to_instruction::<ExecuteInstruction, _, _>(
+            &mut offchain_execute_instruction,
+            mock_fetch_account_data_fn,
+            &MOCK_EXTRA_METAS_STATE,
+        )
+        .await
+        .unwrap();
+
+        // Finally, use the onchain function to add the extra account metas to
+        // the _execute_ CPI instruction from onchain
+        let mut onchain_execute_cpi_instruction = spl_transfer_hook_interface::instruction::execute(
+            &transfer_hook_program_id,
+            &source_pubkey,
+            &mint_pubkey,
+            &destination_pubkey,
+            &authority_pubkey,
+            &validate_state_pubkey,
+            amount,
+        );
+        let mut onchain_execute_cpi_account_infos = vec![
+            source_account_info.clone(),
+            mint_account_info.clone(),
+            destination_account_info.clone(),
+            authority_account_info.clone(),
+            validate_state_account_info.clone(),
+        ];
+        let all_account_infos = &[
+            source_account_info.clone(),
+            mint_account_info.clone(),
+            destination_account_info.clone(),
+            authority_account_info.clone(),
+            validate_state_account_info.clone(),
+            extra_meta_1_account_info.clone(),
+            extra_meta_2_account_info.clone(),
+            extra_meta_3_account_info.clone(),
+            extra_meta_4_account_info.clone(),
+            extra_meta_5_account_info.clone(),
+            extra_meta_6_account_info.clone(),
+        ];
+
+        ExtraAccountMetaList::add_to_cpi_instruction::<ExecuteInstruction>(
+            &mut onchain_execute_cpi_instruction,
+            &mut onchain_execute_cpi_account_infos,
+            &MOCK_EXTRA_METAS_STATE,
+            all_account_infos,
+        )
+        .unwrap();
+
+        // The two `Execute` instructions should have the same accounts
+        assert_eq!(
+            offchain_execute_instruction.accounts,
+            onchain_execute_cpi_instruction.accounts,
+        );
+
+        // Still, the transfer instruction is going to be missing the
+        // the validation account at index 4
+        assert_ne!(
+            offchain_transfer_instruction.accounts,
+            offchain_execute_instruction.accounts,
+        );
+        assert_ne!(
+            offchain_transfer_instruction.accounts[4].pubkey,
+            validate_state_pubkey,
+        );
```

Comment:
I think I'm missing something -- isn't that a problem? The transfer instruction must have *all* required accounts, including the validation account, right?

Reply:
The validation account should be at the 2nd last index. Sorry, I didn't articulate that in the tests, but it's in there along with the transfer hook meta.

Reply:
I'll add this in.